recordCollection.go

```
@@ -0,0 +1,196 @@
+
+		proxied := ""
+		if r.Proxied {
+			proxied = " ; PROXIED"
+		}
+
+		fmt.Fprintf(w, "%s %d %-8s %s%s\n", name, r.TTL, "IN "+r.Type, r.Content, proxied)
+	}
+}
+
+// parseZone will parse a BIND style zone file and return the zone name and
+// a recordCollection.
+func parseZone(r io.Reader) (string, recordCollection, error) {
+	var zoneName string
+	records := recordCollection{}
+
+	for t := range dns.ParseZone(r, "", "") {
+		if t.Error != nil {
+			return "", recordCollection{}, t.Error
+		}
+
+		// Search for zonename while we're at it.
+		soa, found := t.RR.(*dns.SOA)
+		if found {
+			zoneName = strings.Trim(soa.Header().Name, ".")
+		}
+
+		r := newRecord(t)
+		if r != nil {
+			records = append(records, *r)
+		}
+	}
+
+	if zoneName == "" {
+		return "", recordCollection{}, errors.New("Zone name not found")
+	}
+
+	return zoneName, records, nil
+}
+
+// newRecord will instantiate a new cloudflare-compatible DNS record based on
+// a token from our zone file parser.
+func newRecord(in *dns.Token) *cloudflare.DNSRecord {
+	record := &cloudflare.DNSRecord{
+		Name: strings.Trim(in.Header().Name, "."),
+		TTL:  int(in.Header().Ttl),
+	}
+
+	if record.TTL == 1 {
+		record.Proxied = true
+	}
+
+	switch in.RR.(type) {
+	case *dns.A:
+		a := in.RR.(*dns.A)
+		record.Content = a.A.String()
+		record.Type = "A"
+		return record
+
+	case *dns.AAAA:
+		a := in.RR.(*dns.AAAA)
+		record.Content = a.AAAA.String()
+		record.Type = "AAAA"
+		return record
+
+	case *dns.CNAME:
+		cname := in.RR.(*dns.CNAME)
+		record.Content = cname.Target
+		record.Type = "CNAME"
+
+		// CloudFlare does not use the "FQDN-dot". We remove it.
+		if strings.HasSuffix(record.Content, ".") {
+			record.Content = record.Content[:len(record.Content)-1]
+		}
+		return record
+
+	case *dns.MX:
+		mx := in.RR.(*dns.MX)
+		record.Content = strings.Trim(mx.Mx, ".")
+		record.Priority = int(mx.Preference)
+		record.Type = "MX"
+		return record
+
+	// We ignore these for now.
```

Shouldn't we rather return an error, so that we dont add stuff to a zone file expecting it to be passed to CF, just to end up being silently ignored.

---

Godo point. Fixed.